### PR TITLE
BitbucketOrganizations implemented

### PR DESCRIPTION
**Pull request by nmonokov**
PR for #978 
- BitbucketOrganizations implemented
- Unit tests updated

### CODE REVIEW

**Comment by charlesmike** [comment]
@nmonokov thank you for your Pull Request. I'll assign someone to review it soon.

**Comment by charlesmike** [comment]
@amihaiemil please review this Pull Request. Deadline (when it should be merged or closed) is ``2021-02-14T16:26:29.562287``.

You should check if the requirements have been implemented (partially or in full), if there are unit tests covering the changes and if the CI build passes. Feel free to reject the PR or ask for changes if it's too big or not clear enough.

Estimation here is ``30 minutes``, that's how much you will be paid. You will be paid even if this PR gets rejected.

File: self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.function.Consumer;
+
+/**
+ * A Bitbucket Provider Organization Repos.
+ *
+ * @author Nikita Monokov (nmonokov@gmail.com)
+ * @version $Id$
+ * @since 0.0.62
+ */
+public class BitbucketOrganizationRepos implements Repos {
```
Comment (amihaiemil) [inline self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java:20]: @nmonokov Can you write a TODO here, to continue implementing this class?

File: self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * Bitbucket's JSON Resources.
+     */
+    private JsonResources resources;
+
+    /**
+     * Storage used by Organization Repo.
+     */
+    private Storage storage;
```
Comment (amihaiemil) [inline self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java:40]: @nmonokov All attributes of all classes should be final (immutable objects only).

File: self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+     * Ctor.
+     * @param uri Organization Repos URI.
+     * @param owner Current authenticated User.
+     * @param resources Bitbucket's JSON Resources.
+     * @param storage Storage used by Organization Repo.
+     */
+    public BitbucketOrganizationRepos(URI uri,
+                                      User owner,
+                                      JsonResources resources,
+                                      Storage storage) {
```
Comment (amihaiemil) [inline self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java:52]: @nmonokov method and constructor parameters final as well please

File: self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizations.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import com.selfxdsd.api.storage.Storage;
+
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.util.Iterator;
+
+/**
+ * A Bitbucket Provider Organizations.
```
Comment (amihaiemil) [inline self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizations.java:15]: @nmonokov "A User's Organizations on Bitbucket"

File: self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganization.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import com.selfxdsd.api.Organization;
+import com.selfxdsd.api.Repos;
+import com.selfxdsd.api.User;
+import com.selfxdsd.api.storage.Storage;
+
+import javax.json.JsonObject;
+import java.net.URI;
+
+/**
+ * A Bitbucket Provider Organization.
```
Comment (amihaiemil) [inline self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganization.java:12]: @nmonokov "A Bitbucket Organization (Team)"

**Review by amihaiemil — CHANGES_REQUESTED** [review]
@nmonokov Just a few comments.

Also, this PR is quite big (estimation of the ticket was only 60min). Orangizations were enough, you didn't have to also implement the Org Repos.  

Read this, about estimations, microtasking and TODOs: https://docs.self-xdsd.com/microtasking.html

**Comment by amihaiemil** [comment]
@nmonokov also, Travis build is failing. Checkstyle has some complaints: https://travis-ci.org/github/self-xdsd/self-core/builds/758534706#L1688

File: self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java (hunk 1 above)
Comment (nmonokov) [reply self-core-impl/src/main/java/com/selfxdsd/core/BitbucketOrganizationRepos.java:20]: I left it with unsupported exceptions. I couldn't understand where is the line between my task and repos and just made this class with defaults :)

**Comment by nmonokov** [comment]
@amihaiemil Travis seems to fail yet another time. Can you help me understand how to pass `Method 'iterator' is not designed for extension - needs to be abstract, final or empty. [DesignForExtension]` check?

**Comment by amihaiemil** [comment]
@nmonokov Yes, that error occurs because your class is not final.

There is an architectural rule in Self XDSD: any class must be either ``final`` or ``abstract``. This is to prevent extention of non-abstract classes (which usually causes spaghetti code really fast). 

So, make sure that any class you introduce is final (``public final class ...``).

**Comment by amihaiemil** [comment]
@nmonokov I'll merge this now and fix it quick, no problem.

**Comment by amihaiemil** [comment]
@nmonokov 

> I left it with unsupported exceptions. I couldn't understand where is the line between my task and repos and just made this class with defaults :)

Any time you leave an unsupported exception (not yet implemented) you should also write a TODO (which will be transformed into a new Issue automatically), so the development can continue.

> where is the line between my task and repos

You should only work for the task's estimation (usually 60min). Anything else, write TODOs.

**Comment by amihaiemil** [comment]
@nmonokov I'm not sure if you noticed, but ``charlesmike`` is a chatbot, not a real person. That's what Self XDSD is: a platform for automated project management.

**Comment by charlesmike** [comment]
@amihaiemil thank you for resolving this ticket. I've just added it to your active invoice. You can always check all your invoices and more on the [Contributor Dashboard](https://self-xdsd.com).

**Comment by nmonokov** [comment]
@amihaiemil yeah, I've noticed that `charlesmike` is not a person. Though it was a frustrating experience at first :D. As for todos and time limit, eventually i'll get a grasp on this concept. This is really unusual one.

**Pull request merged**